model/model_training/custom_datasets/formatting.py

```diff
@@ -15,20 +28,291 @@ def format_system_prefix(prefix, eos_token):
+    WA = "wa"  # Walloon
+    CY = "cy"  # Welsh
+    WO = "wo"  # Wolof
+    FY = "fy"  # Western Frisian
+    XH = "xh"  # Xhosa
+    YI = "yi"  # Yiddish
+    YO = "yo"  # Yoruba
+    ZA = "za"  # Zhuang, Chuang
+    ZU = "zu"  # Zulu
+
+
+class DatasetEntry(BaseModel):
+    questions: list[str]
+    answers: list[str]
+    context: str | None
+    lang: Language | None
+    length: int | None
+    quality: float | None
+    humor: float | None
+    creativity: float | None
+
+    @validator("length")
+    def above_zero(cls, v) -> int:
+        if v is not None and v < 0:
+            raise ValueError(f"Length cannot be lower than 0. Received {v}")
+        return v
+
+    @validator("quality", "humor", "creativity")
+    def between_0_1(cls, v, field: ModelField) -> float:
+        if v is not None and not (0 <= v <= 1):
+            raise ValueError(f"Field {field.name} must be between 0 and 1. Received {v}.")
+        return v
+
+    def system_tag(self, eos_token: str) -> str | None:
+        relevant_system_infos = [
+            (k, v)
+            for k, v in self.__dict__.items()
+            if k not in ["questions", "answers"] and v is not None and v.replace("\n", "")
+        ]
+        if len(relevant_system_infos) > 0:
+            shuffle(relevant_system_infos)
+            system_tag_key_values = "\n".join([f"{k}: {v}" for k, v in relevant_system_infos])
+            system_tag = f"{QA_SPECIAL_TOKENS['System']}{system_tag_key_values}\n{eos_token}"
+            return system_tag
+
+    def get_formatted(self, mode: Mode, eos_token: str) -> str | list[str]:
+        system_tag = self.system_tag(eos_token)
+        if mode == Mode.rl:
+            if system_tag is not None:
+                return f"{system_tag}{QA_SPECIAL_TOKENS['Question']}{self.questions[0]}{QA_SPECIAL_TOKENS['Answer']}"
+            else:
+                return f"{QA_SPECIAL_TOKENS['Question']}{self.questions[0]}{QA_SPECIAL_TOKENS['Answer']}"
+        if system_tag is not None:
+            qa_list = [system_tag]
+        else:
+            qa_list = list()
+        for q, a in zip_longest(self.questions, self.answers):
+            match (q, a):
+                case (str(), str()):
+                    qa_list.extend(
+                        [
+                            f"{QA_SPECIAL_TOKENS['Question']}{q}{eos_token}",
+                            f"{QA_SPECIAL_TOKENS['Answer']}{a}{eos_token}",
+                        ]
+                    )
+                case (str(), None):
+                    qa_list.append(f"{QA_SPECIAL_TOKENS['Question']}{q}{eos_token}")
+                case (None, None):
+                    break
+                case (None, str()):
+                    raise ValueError("Received answer without getting corresponding question. Aborting")
+        if mode == Mode.sft:
+            return qa_list
+        elif mode == Mode.rm:
+            raise NotImplementedError("This is currently not implemented.")
```

Comment:
Does this mean that RM training will not work after this is merged?

Reply:
Yep, that is correct. Currently it is only implemented for Dolly, which does not support rm, so no change for now. But this needs to be implemented soon.